sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java

```diff
@@ -22,69 +22,54 @@
-        final String[] sourceIndexName = request.sourceIndexName();
-        final String[] targetIndexName = request.targetIndexName();
+    public ClusterState execute(ClusterState currentState, BulkRenameIndexRequest request) throws Exception {
 
-        logger.info("rename indexes '[{}]' to '[{}]'", sourceIndexName, targetIndexName);
+        final String[] sourceIndexNames = request.sourceIndices();
+        final String[] targetIndexNames = request.targetIndices();
 
+        logger.info("rename indexes '[{}]' to '[{}]'", sourceIndexNames, targetIndexNames);
         MetaData.Builder mdBuilder = null;
         ClusterBlocks.Builder blocksBuilder = ClusterBlocks.builder()
             .blocks(currentState.blocks());
 
-        Index[] sourceIndices = indexNameExpressionResolver.concreteIndices(currentState,
-            STRICT_INDICES_OPTIONS, sourceIndexName);
-        Index[] sourceIndicesOrdered = new Index[sourceIndexName.length];
-
-        // ensure the order in input is respected
-        for (int i = 0; i < sourceIndexName.length; i++) {
-            for (int j = 0; j < sourceIndices.length; j++) {
-                if (sourceIndexName[i].equals(sourceIndices[j].getName())) {
-                    sourceIndicesOrdered[i] = sourceIndices[j];
-                    break;
-                }
-            }
-        }
-
         MetaData metaData = currentState.getMetaData();
         mdBuilder = MetaData.builder(metaData);
 
-        renameIndices(metaData, mdBuilder, blocksBuilder, sourceIndicesOrdered, targetIndexName);
+        renameIndices(metaData, mdBuilder, blocksBuilder, sourceIndexNames, targetIndexNames);
+        return ClusterState.builder(currentState).metaData(mdBuilder).blocks(blocksBuilder).build();
+    }
 
-        // The MetaData will always be overridden (and not merged!) when applying it on a cluster state builder.
-        // So we must re-build the state with the latest modifications before we pass this state to possible modifiers
-        // or return it.
-        // Otherwise they would operate on the old MetaData and would just ignore any modifications.
-        currentState = ClusterState.builder(currentState).metaData(mdBuilder).blocks(blocksBuilder).build();
-        return currentState;
+    static void renameIndices(MetaData metaData,
+                              MetaData.Builder mdBuilder,
+                              ClusterBlocks.Builder blocksBuilder,
+                              String[] sourceIndexNames,
+                              String[] targetIndexNames) {
+        for (int i = 0; i < sourceIndexNames.length; i++) {
+            String sourceIndex = sourceIndexNames[i];
+            IndexMetaData indexMetaData = metaData.index(sourceIndex);
+            IndexMetaData targetIndexMetadata = IndexMetaData.builder(indexMetaData)
+                .index(targetIndexNames[i]).build();
+            mdBuilder.remove(sourceIndex);
```

Comment:
I think by creating the new metadata with all indices present and then mutating it we still have an order dependency here.

E.g.

```
rename actions:
  x -> y
  y -> z

results in:

mdBuilder.remove(x)
mdBuilder.put(mdForY)
mdBuilder.remove(y)
mdBuilder.put(mdForZ)
```